contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol

```diff
@@ -0,0 +1,229 @@
+
+  bytes private s_donPublicKey;
+  mapping(address => bytes) private s_nodePublicKeys;
+
+  constructor(
+    address router,
+    bytes memory config,
+    address linkToNativeFeed
+  ) OCR2Base(true) FunctionsBilling(router, config, linkToNativeFeed) {}
+
+  /**
+   * @inheritdoc ITypeAndVersion
+   */
+  function typeAndVersion() public pure override returns (string memory) {
+    return "Functions Coordinator v1";
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function getDONPublicKey() external view override returns (bytes memory) {
+    return s_donPublicKey;
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function setDONPublicKey(bytes calldata donPublicKey) external override onlyRouterOwner {
+    if (donPublicKey.length == 0) {
+      revert EmptyPublicKey();
+    }
+    s_donPublicKey = donPublicKey;
+  }
+
+  /**
+   * @dev check if node is in current transmitter list
+   */
+  function _isTransmitter(address node) internal view returns (bool) {
+    address[] memory nodes = this.transmitters();
+    for (uint256 i = 0; i < nodes.length; i++) {
+      if (nodes[i] == node) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function setNodePublicKey(address node, bytes calldata publicKey) external override {
+    // Owner can set anything. Transmitters can set only their own key.
+    if (!(msg.sender == IOwnable(address(s_router)).owner() || (_isTransmitter(msg.sender) && msg.sender == node))) {
+      revert UnauthorizedPublicKeyChange();
+    }
+    s_nodePublicKeys[node] = publicKey;
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function deleteNodePublicKey(address node) external override {
+    // Owner can delete anything. Others can delete only their own key.
+    if (!(msg.sender == IOwnable(address(s_router)).owner() || msg.sender == node)) {
+      revert UnauthorizedPublicKeyChange();
+    }
+    delete s_nodePublicKeys[node];
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function getAllNodePublicKeys() external view override returns (address[] memory, bytes[] memory) {
+    address[] memory nodes = this.transmitters();
+    bytes[] memory keys = new bytes[](nodes.length);
+    for (uint256 i = 0; i < nodes.length; i++) {
+      keys[i] = s_nodePublicKeys[nodes[i]];
+    }
+    return (nodes, keys);
+  }
+
+  /**
+   * @inheritdoc IFunctionsCoordinator
+   */
+  function sendRequest(
+    uint64 subscriptionId,
+    bytes calldata data,
+    uint32 gasLimit,
+    address caller,
+    address subscriptionOwner
+  )
+    external
+    override
+    onlyRouter
+    returns (
+      bytes32 requestId,
+      uint96 estimatedCost,
+      uint256 gasAfterPaymentCalculation,
+      uint256 requestTimeoutSeconds
+    )
+  {
+    {
+      if (data.length == 0) {
+        revert EmptyRequestData();
+      }
+
+      (uint16 version, ) = Functions.decodeRequest(data);
```

Comment:
Decode the whole request just to get the version? Seems too expensive.

Reply:
We could just not encode it :shrug:

Reply:
Added a function to the Client contract for passing bytes directly. This is also useful for when we update the request data.

Change for removing this double coding still incoming.